server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java

```diff
@@ -0,0 +1,207 @@
+ * </ol>
+ * </li>
+ * </ol>
+ * <h1>Creating a Snapshot</h1>
+ *
+ * <p>Creating a snapshot in the repository happens in the three steps described in detail below.</p>
+ *
+ * <h2>Initializing a Snapshot in the Repository</h2>
+ *
+ * <p>Creating a snapshot in the repository starts with a call to {@link org.elasticsearch.repositories.Repository#initializeSnapshot} which
+ * the blob store repository implements via the following actions:</p>
+ * <ol>
+ * <li>Verify that no snapshot by the requested name exists.</li>
+ * <li>Write a blob containing the cluster metadata to the root of the blob store repository at {@code /meta-${snapshot-uuid}.dat}</li>
+ * <li>Write the metadata for each index to a blob in that index's directory at
+ * {@code /indices/${index-snapshot-uuid}/meta-${snapshot-uuid}.dat}</li>
+ * </ol>
+ * TODO: This behavior is problematic, adjust these docs once https://github.com/elastic/elasticsearch/issues/41581 is fixed
+ *
+ * <h2>Writing Shard Data (Segments)</h2>
+ *
+ * <p>Once all the metadata has been written by the snapshot initialization, the snapshot process moves on to writing the actual shard data
+ * to the repository by invoking {@link org.elasticsearch.repositories.Repository#snapshotShard} on the data-nodes that hold the primaries
+ * for the shards in the current snapshot. It is implemented as follows:</p>
+ *
+ * <p>Note:</p>
+ * <ul>
+ * <li>For each shard {@code i} in a given index, its path in the blob store is located at {@code /indices/${index-snapshot-uuid}/${i}}</li>
+ * <li>All the following steps are executed exclusively on the shard's primary's data node.</li>
+ * </ul>
+ *
+ * <ol>
+ * <li>Create the {@link org.apache.lucene.index.IndexCommit} for the shard to snapshot.</li>
+ * <li>List all blobs in the shard's path. Find the {@link org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshots} blob
+ * with name {@code index-${N}} for the highest possible value of {@code N} in the list to get the information of what segment files are
+ * already available in the blobstore.</li>
+ * <li>By comparing the files in the {@code IndexCommit} and the available file list from the previous step, determine the segment files
+ * that need to be written to the blob store. For each segment that needs to be added to the blob store, generate a unique name by combining
+ * the segment data blob prefix {@code __} and a UUID and write the segment to the blobstore.</li>
+ * <li>After completing all segment writes, a blob containing a
+ * {@link org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot} with name {@code snap-${snapshot-uuid}.dat} is written to
+ * the shard's path and contains a list of all the files referenced by the snapshot as well as some metadata about the snapshot. See the
+ * documentation of {@code BlobStoreIndexShardSnapshot} for details on its contents.</li>
+ * <li>Once all the segments and the {@code BlobStoreIndexShardSnapshot} blob have been written, an updated
```

Comment:
Can we elaborate here why we need both BlobStoreIndexShardSnapshot and BlobStoreIndexShardSnapshots? AFAIK, it's needed exclusively for delete snapshot operations.

Reply:
We currently use `BlobStoreIndexShardSnapshots` to determine the segments to upload when creating incremental snapshots of a shard as well as what to delete.
Technically speaking `BlobStoreIndexShardSnapshots` is redundant as you could rebuild it from all the `BlobStoreIndexShardSnapshot` blobs. In practice this might get tricky since it implies downloading potentially hundreds of them (you have exactly as many as you have snapshots referencing the shard).
I think I'd really just leave the details of what is in `BlobStoreIndexShardSnapshot` precisely to the link to that class (it contains some additional metadata about the time it took to snapshot the shard that we expose via an API). It seems to me it's easier to read this by not mentioning the duplication of data here, that just has the reader wondering why (and it's a fair question :D) when understanding that the `BlobStoreIndexShardSnapshots` holds the list of segments is all that really matters for the big picture here?

Reply:
Ok